pkg/controller/mizar/mizar-starter-controller.go

```diff
@@ -0,0 +1,165 @@
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().EventsWithMultiTenancy(metav1.NamespaceAll, metav1.TenantAll)})
+
+	c := &MizarStarterController{
+		kubeClient:        kubeClient,
+		lister:            informer.Lister(),
+		listerSynced:      informer.Informer().HasSynced,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerForMizarStarter),
+		controllerContext: controllerContext,
+		startHandler:      startHandler,
+	}
+
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.createObj,
+	})
+	c.lister = informer.Lister()
+	c.listerSynced = informer.Informer().HasSynced
+
+	c.handler = c.handle
+
+	return c
+}
+
+// Run begins watching and handling.
+func (c *MizarStarterController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %v controller", controllerForMizarStarter)
+	defer klog.Infof("Shutting down %v controller", controllerForMizarStarter)
+
+	if !controller.WaitForCacheSync(controllerForMizarStarter, stopCh, c.listerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *MizarStarterController) createObj(obj interface{}) {
+	key, _ := controller.KeyFunc(obj)
+	c.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the handler is never invoked concurrently with the same key.
+func (c *MizarStarterController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MizarStarterController) processNextWorkItem() bool {
+	workItem, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+
+	key := workItem.(string)
+	defer c.queue.Done(key)
+
+	err := c.handler(key)
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Handle %v of key %v failed with %v", controllerForMizarStarter, key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *MizarStarterController) handle(key string) error {
+	klog.Infof("Entering handling for %v. key %s", controllerForMizarStarter, key)
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished handling %v %q (%v)", controllerForMizarStarter, key, time.Since(startTime))
+	}()
+
+	tenant, namespace, name, err := cache.SplitMetaTenantNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, err := c.lister.ConfigMapsWithMultiTenancy(namespace, tenant).Get(name)
+	if err != nil {
+		return err
+	}
+
+	klog.V(4).Infof("Handling %v %s/%s/%s hashkey %v", controllerForMizarStarter, tenant, namespace, obj.Name, obj.HashKey)
+
+	if key == "default/mizar-grpc-service" {
```

Comment:
Don't check the key directly. Instead check the value of tenant/ns/name. The key format may be changed in the future.